sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java

```diff
@@ -0,0 +1,152 @@
+ * {@link #sampleUpdateMs}.
+ */
+public class MovingFunction {
+  /**
+   * How far back to retain samples, in ms.
+   */
+  private final long samplePeriodMs;
+
+  /**
+   * How frequently to update the moving function, in ms.
+   */
+  private final long sampleUpdateMs;
+
+  /**
+   * How many buckets are considered 'significant'?
+   */
+  private final int numSignificantBuckets;
+
+  /**
+   * How many samples are considered 'significant'?
+   */
+  private final int numSignificantSamples;
+
+  /**
+   * Function for combining sample values.
+   */
+  private final Combine.BinaryCombineLongFn function;
+
+  /**
+   * Minimum/maximum/sum of all values per bucket.
+   */
+  private final long[] buckets;
+
+  /**
+   * How many samples have been added to each bucket.
+   */
+  private final int[] numSamples;
+
+  /**
+   * Time of start of current bucket.
+   */
+  private long currentMsSinceEpoch;
+
+  /**
+   * Index of bucket corresponding to above timestamp, or -1 if no entries.
+   */
+  private int currentIndex;
+
+  public MovingFunction(long samplePeriodMs, long sampleUpdateMs,
+                        int numSignificantBuckets, int numSignificantSamples,
+                        Combine.BinaryCombineLongFn function) {
+    this.samplePeriodMs = samplePeriodMs;
+    this.sampleUpdateMs = sampleUpdateMs;
+    this.numSignificantBuckets = numSignificantBuckets;
+    this.numSignificantSamples = numSignificantSamples;
+    this.function = function;
+    int n = (int) (samplePeriodMs / sampleUpdateMs);
+    buckets = new long[n];
+    Arrays.fill(buckets, function.identity());
+    numSamples = new int[n];
+    Arrays.fill(numSamples, 0);
+    currentMsSinceEpoch = -1;
+    currentIndex = -1;
+  }
+
+  /**
+   * Flush stale values.
+   */
+  private void flush(long nowMsSinceEpoch) {
+    Preconditions.checkArgument(nowMsSinceEpoch >= 0);
+    if (currentIndex < 0) {
+      currentMsSinceEpoch = nowMsSinceEpoch - (nowMsSinceEpoch % sampleUpdateMs);
+      currentIndex = 0;
+    }
+    Preconditions.checkArgument(nowMsSinceEpoch >= currentMsSinceEpoch);
+    int newBuckets =
+        Math.min((int) ((nowMsSinceEpoch - currentMsSinceEpoch) / sampleUpdateMs),
+                 buckets.length);
+    while (newBuckets > 0) {
+      currentIndex = (currentIndex + 1) % buckets.length;
+      buckets[currentIndex] = function.identity();
+      numSamples[currentIndex] = 0;
+      newBuckets--;
+      currentMsSinceEpoch += sampleUpdateMs;
+    }
+  }
+
+  /**
+   * Add {@code value} at {@code nowMsSinceEpoch}.
+   */
+  public void add(long nowMsSinceEpoch, long value) {
+    flush(nowMsSinceEpoch);
+    buckets[currentIndex] = function.apply(buckets[currentIndex], value);
+    numSamples[currentIndex]++;
+  }
+
+  /**
+   * Return the minimum/maximum/sum of all retained values within {@link #samplePeriodMs}
+   * of {@code nowMsSinceEpoch}.
+   */
+  public long get(long nowMsSinceEpoch) {
```

Comment:
Don't want to pass Clock since I snapshot 'now' per logical unit of work.
Don't want to rename since I like to spell out exact units.
(And don't want to use Instant since it is surprisingly expensive.)